### PR TITLE
[MRG] Shelving for NiftiMasker

### PR DESCRIPTION
**Pull request by arthurmensch**
Needs code + doc. @aabraham does it look lightweight enough ?

### CODE REVIEW

**Comment by AlexandreAbraham** [comment]
To, the simplest option would be to add a `shelve` parameter to `masker._cache` which would prevent you from ugly stuff like passing `self` as an argument (which is weird... If you need to pass self as an argument, just create a method).

**Comment by arthurmensch** [comment]
Address issue #203

**Comment by arthurmensch** [comment]
Ready for review pin @AlexandreAbraham

File: nilearn/input_data/tests/test_multi_nifti_masker.py [hunk 1]
```diff
@@ -127,3 +130,32 @@ def test_joblib_cache():
...
+            assert_array_equal(epi_shelved, epi)
+
+        epi = masker.fit_transform(epi_img1)
+        epi_shelved = masker_shelved.fit_transform(epi_img1)
+        epi_shelved = epi_shelved.get()
+        assert_array_equal(epi_shelved, epi)
+    finally:
+        # enables to delete "filename" on windows
+        del masker
+        shutil.rmtree(cachedir, ignore_errors=True)
```
Comment (arthurmensch) [inline nilearn/input_data/tests/test_multi_nifti_masker.py]: Actually there is an end of line in the file, Github bug ?

File: nilearn/_utils/cache_mixin.py [hunk 2]
```diff
@@ -189,6 +206,10 @@ def _cache(self, func, func_memory_level=1, **kwargs):
             The memory_level from which caching must be enabled for the wrapped
             function.
 
+        shelve: boolearn,
```
Comment (AlexandreAbraham) [inline nilearn/_utils/cache_mixin.py]: nilearn has its own boolean type, the boolearn :P

File: nilearn/input_data/multi_nifti_masker.py [hunk 3]
```diff
@@ -93,6 +93,11 @@ class MultiNiftiMasker(NiftiMasker, CacheMixin):
         The number of CPUs to use to do the computation. -1 means
         'all CPUs', -2 'all CPUs but one', and so on.
 
+    shelve: boolearn,
```
Comment (AlexandreAbraham) [inline nilearn/input_data/multi_nifti_masker.py]: booleaaaaaaarn

File: nilearn/input_data/multi_nifti_masker.py [hunk 4]
```diff
@@ -93,6 +93,11 @@ class MultiNiftiMasker(NiftiMasker, CacheMixin):
...
+        Fix behavior of the transform method : whether to return a
+        MemorizedResult, callable by a .get() method,
+        instead of the return value of func
```
Comment (AlexandreAbraham) [inline nilearn/input_data/multi_nifti_masker.py]: I'm afraid that this doc is highly not understand by non technical people.

File: nilearn/input_data/multi_nifti_masker.py [hunk 5]
```diff
@@ -175,23 +183,24 @@ def fit(self, imgs=None, y=None):
...
             self.mask_img_ = self._cache(compute_mask,
-                        ignore=['n_jobs', 'verbose', 'memory'])(
-                            imgs,
-                            target_affine=self.target_affine,
-                            target_shape=self.target_shape,
-                            n_jobs=self.n_jobs,
-                            memory=self.memory,
-                            verbose=max(0, self.verbose - 1),
-                        **mask_args)
+                                         ignore=['n_jobs', 'verbose',
```
Comment (AlexandreAbraham) [inline nilearn/input_data/multi_nifti_masker.py]: can you put `compute_mask, ignore=[...])` on  a single line for prettier code?

**Comment by AlexandreAbraham** [comment]
I am +1 given some small comments. The next step will be an integration of that in CanICA and the addition of an example.

**Comment by arthurmensch** [comment]
What should we use as an example ? Simple plotting function ?

**Comment by GaelVaroquaux** [comment]
> What should we use as an example ? Simple plotting function ?

No need for example. Just integration in CanICA and DictLearning.

**Comment by arthurmensch** [comment]
Unless I am allow to rewrite the pipeline as:
- mask and shelve all data in parallel
- reduce all shelved data in parallel

Instead of mask and reduce all data in parallel without shelving (which is what we currently do)

**Comment by AlexandreAbraham** [comment]
Yeah, now that you point it out, it is meaningless in CanICA.

**Comment by arthurmensch** [comment]
Unless we allow dimension reduction within `MultiNiftiMasker`, which is something that I advocated for once upon a time but may not be such a good idea.

On the other hand, the rewriting of the pipeline does make sense to me as it would simplifies the code quite much -- but it would be a large PR with no improvement for the end user.

**Comment by GaelVaroquaux** [comment]
> Yeah, now that you point it out, it is meaningless in CanICA.

Yes, indeed, because the reduction does not happen in the NiftiMasker,
and that's the operation done in Parallel. So that PR is actually only
useful for our internal needs, isn't it?

**Comment by arthurmensch** [comment]
We also need to disable this feature for `sklearn <= 0.14.1`

**Comment by arthurmensch** [comment]
It is. It would be useful for MSDL if it ever comes to nilearn, and for online (sparse) PCA if it ever comes to nilearn. Wonder if it would not be useful in decoding as well @dohmatob ?

**Comment by GaelVaroquaux** [comment]
> It is. It would be useful for MSDL if it ever comes to nilearn, and for online
> (sparse) PCA if it ever comes to nilearn.

Indeed. The problem is that those are still pending.

> Wonder if it would not be useful in decoding as well @dohmatob ?

Maybe. But I am not sure that it would be the case in the SpaceNet
object.

**Comment by arthurmensch** [comment]
I might be wrong, but I don't think `MultiNiftiMasker` if actually used outside of `decomposition` module isn't it ?

**Comment by AlexandreAbraham** [comment]
It's actually not really used in decomposition, we just use the fit to compute the mask.

**Comment by arthurmensch** [comment]
Should I have a shot at decomposition refactorisation or should we just advertise in a small example ? Btw there are estimators in the code (SearchLight, SpaceNet), that uses a NiftiMasker on list of images, while they could be parallelized.

**Comment by GaelVaroquaux** [comment]
> Should I have a shot at decomposition refactorisation or should we just
> advertise in a small example ?

No refactor. An example in the advanced section would be good. Not sure
about what, though :$

> Btw there are estimators in the code (SearchLight, SpaceNet), that uses
> a NiftiMasker on list of images, while they could be parallelized.

Good point. Could you add an issue where you expose the point.

**Comment by AlexandreAbraham** [comment]
SearchLight does not iterate on a list of images.

**Comment by arthurmensch** [comment]
Just added an issue on SpaceNet: it actually fails on list of images.

**Comment by AlexandreAbraham** [comment]
@arthurmensch, I'm working on integrating MSDL in nilearn and I thinking that providing shelving would really be a big plus. I think that, as suggested by Gael, we should make this option public for the moment and just use it internally. Basically, a few modifications are needed in your PR. Do you want to take care of it or do you want me to take over the PR?

**Comment by arthurmensch** [comment]
I'll take it!
Le 26 mai 2016 16:33, "Alexandre Abraham" <notifications@github.com> a
écrit :

**Comment by AlexandreAbraham** [comment]
Sorry, I meant "not public". So, I think that you just need to rebase and then removing the option from MultiNiftiMasker constructor. Maybe we should create a method `_enable shelving` or so? I leave that to you :innocent:

**Comment by AlexandreAbraham** [comment]
Good! Thanks.

**Comment by AlexandreAbraham** [comment]
I take over the PR to fix the failing tests.

**Comment by arthurmensch** [comment]
Please do, I am sorry not be available today.
Le 9 juin 2016 10:15, "Alexandre Abraham" <notifications@github.com> a
écrit :

**Comment by AlexandreAbraham** [comment]
Replaced by #1114

**Comment by AlexandreAbraham** [comment]
> Please do, I am sorry not be available today.

Not a problem! It's just a matter of changing 3 lines ;).
